grafana/grafana
# Loki: Fix support of ad-hoc filters for specific queries

**What this PR does / why we need it**:
In this PR, we are changing a way how we add ad-hoc filters. While previously we've used complex handwritten regex matching code which did not support some queries (metric Loki query with parser). It was also buggy in some cases (see issue https://github.com/grafana/grafana/issues/47701). 

This PR removes the code and instead uses recently introduced LogQL parser to simplify this. We are using similar logic as is used in [Prometheus](public/app/plugins/datasource/prometheus/add_label_to_query.ts). 

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/47701

**Special notes for your reviewer**:

To test this PR:
1. Start loki `make devenv sources=loki`
2. Create new dashboard
3. Go to `Add variables` -> Choose `Ad hoc filters` -> `gdev-loki` data source
<img width="957" alt="image" src="https://user-images.githubusercontent.com/30407135/175004012-eca07eba-3acf-4c58-99da-4569f387ae56.png">
4. Add ad hoc filter
<img width="314" alt="image" src="https://user-images.githubusercontent.com/30407135/175011110-1e131b48-9aab-49bc-af37-780aca987552.png">
5. Add new panel -> Choose `gdev-loki` data source
6. Run queries with open network tab to see if label-value from ad hoc filter was correctly added
<img width="1904" alt="image" src="https://user-images.githubusercontent.com/30407135/175004381-e74c3348-0b44-45f4-9b6e-742c082850ab.png">

Examples of queries to run: 
- `{source="data"}` => `{source="data", place="luna"}`
- `{source="data", place="luna" }`  => `{source="data", place="luna"}`
- `{source="data"} | logfmt`  => `{source="data"} | logfmt | place="luna"}`
- `rate({source="data"} | logfmt [5m])`  => `rate({source="data"} | logfmt | place="luna" [5m])`

Here are bunch of other queries examples that you can use: https://github.com/grafana/lezer-logql/blob/main/test/expression.txt
